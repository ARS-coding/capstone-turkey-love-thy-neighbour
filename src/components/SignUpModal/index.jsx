import React from "react";

import { Modal, Button, Card } from "react-bootstrap";

import { useFormik } from "formik";

import { useDispatch, useSelector } from "react-redux";

import { useHistory } from "react-router-dom";

import { removeOneProp, setUserDocument } from "../../hooks/index";

import { auth } from "../../firebaseConfig";

import "./index.scss";
import logo from "../../images/logo.svg";
import {
  SignInUpButton,
  SignInUpGoogleButton,
  SignInUpFacebookButton,
} from "../CustomButtons";

const SignUpModal = () => {
  const dispatch = useDispatch();
  const isSignUpOpen = useSelector((state) => state.popup.isSignUpOpen);
  const history = useHistory();

  const validate = (values) => {
    const errors = {};
    if (!values.firstName) {
      errors.firstName = "Required";
    }
    if (!values.lastName) {
      errors.lastName = "Required";
    }
    if (!values.email) {
      errors.email = "Required";
    }
    if (!values.password) {
      errors.password = "Required";
    }
    if (!values.repeatedPassword) {
      errors.repeatedPassword = "Required";
    }
    return errors;
  };

  const formik = useFormik({
    initialValues: {
      firstName: "",
      lastName: "",
      email: "",
      password: "",
      repeatedPassword: "",
    },
    validate,
    onSubmit: (values, { resetForm, setSubmitting }) => {
      resetForm();
      const objWithoutPasswordConfigProp = removeOneProp(
        values,
        "repeatedPassword"
      );
      auth
        .createUserWithEmailAndPassword(values.email, values.password)
        .then((userCred) => {
          setUserDocument(userCred.user.uid, objWithoutPasswordConfigProp);
          return userCred;
        }) // set the document in firestore
        .then((userCred) => {
          dispatch({ type: "signUp" });
          history.push(`/profile/${userCred.user.uid}`);
          dispatch({ type: "editProfile" });
        }) // take the user to their profile
        .catch((error) =>
          console.error(
            "A problem occured while your account being created!",
            error
          )
        );
      setSubmitting(false);
    },
  });

  return (
    <Modal
      show={isSignUpOpen}
      onHide={() => {
        dispatch({ type: "signUp" });
      }}
      id="sign-up-modal"
    >
      <Modal.Header>
        <img src={logo} alt="logo" />
        <h2>Sign Up</h2>
        <Button
          type="button"
          data-toggle="modal"
          className="btn-close"
          aria-label="Close"
          onClick={() => {
            dispatch({ type: "signUp" });
          }}
        />
      </Modal.Header>

      <Modal.Body>
        <Card>
          <Card.Body>
            <form
              onSubmit={formik.handleSubmit}
              className="d-flex flex-column"
              id="sign-up-form"
            >
              <input
                className="sign-up-form-item p-2 "
                id="firstName"
                name="firstName"
                placeholder="First Name"
                onChange={formik.handleChange}
                value={formik.values.firstName}
                onBlur={formik.handleBlur}
              />
              {formik.touched.firstName && formik.errors.firstName ? (
                <div className="error-msg">{formik.errors.firstName}</div>
              ) : null}
              <input
                className="sign-up-form-item p-2 "
                id="lastName"
                name="lastName"
                placeholder="Last Name"
                onChange={formik.handleChange}
                value={formik.values.lastName}
                onBlur={formik.handleBlur}
              />
              {formik.touched.lastName && formik.errors.lastName ? (
                <div className="error-msg">{formik.errors.lastName}</div>
              ) : null}
              <input
                className="sign-up-form-item p-2 "
                id="email"
                name="email"
                type="email"
                placeholder="Email"
                onChange={formik.handleChange}
                value={formik.values.email}
                onBlur={formik.handleBlur}
              />
              {formik.touched.email && formik.errors.email ? (
                <div className="error-msg">{formik.errors.email}</div>
              ) : null}
              <input
                className="sign-up-form-item p-2 "
                id="password"
                name="password"
                type="password"
                placeholder="Password"
                onChange={formik.handleChange}
                value={formik.values.password}
                onBlur={formik.handleBlur}
              />
              {formik.touched.password && formik.errors.password ? (
                <div className="error-msg">{formik.errors.password}</div>
              ) : null}
              <input
                className="sign-up-form-item p-2 "
                id="repeatedPassword"
                name="repeatedPassword"
                type="password"
                placeholder="Repeat Password"
                onChange={formik.handleChange}
                value={formik.values.repeatedPassword}
                onBlur={formik.handleBlur}
              />
              {formik.touched.password && formik.errors.password ? (
                <div className="error-msg">{formik.errors.password}</div>
              ) : null}
            </form>
          </Card.Body>
        </Card>
      </Modal.Body>

      <div className="two-footer-wrapper">
        <Modal.Footer className="sign-up-buttons d-flex flex-column align-items-stretch">
          <SignInUpButton
            type="submit"
            disabled={formik.isSubmitting}
            form="sign-up-form"
          >
            Sign Up
          </SignInUpButton>
          <SignInUpGoogleButton type="submit" disabled={formik.isSubmitting}>
            Sign Up With Google
          </SignInUpGoogleButton>
          <SignInUpFacebookButton type="submit" disabled={formik.isSubmitting}>
            Sign Up With Facebook
          </SignInUpFacebookButton>
        </Modal.Footer>
        <Modal.Footer className="d-flex flex-column align-items-center">
          <span>
            Already Got an{" "}
            <a
              href="/"
              onClick={(event) => {
                event.preventDefault();
                dispatch({ type: "signUp" });
                dispatch({ type: "signIn" });
              }}
            >
              Account
            </a>
            ?
          </span>
        </Modal.Footer>
      </div>
    </Modal>
  );
};

export default SignUpModal;
