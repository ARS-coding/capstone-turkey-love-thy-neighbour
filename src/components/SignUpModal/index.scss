@import "../../style/variables";

#sign-up-modal {
  .modal-header {
    h2 {
      margin-left: auto;
      margin-right: auto;
      margin-bottom: 0;
    }

    button {
      margin: 0;
    }
  }

  form {
    input {
      width: 100%;
    }
    .error-msg {
      font-size: 12px;
      color: rgb(122, 12, 22);
      margin-top: 5px;
      margin-bottom: 5px;
      margin-left: 3px;
    }
  }

  .sign-up-buttons {
    padding: 16px;

    button {
      margin-right: 0 !important;
      margin-left: 0 !important;
    }

    button:nth-child(2) {
      margin-top: 5px;
      margin-bottom: 5px;
    }
  }

  .sign-up-form-item {
    margin-top: 10px;
    border: 1px solid $dark-blue;
  }
}
