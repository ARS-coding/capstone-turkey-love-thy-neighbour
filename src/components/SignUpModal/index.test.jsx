// NOT WORKING

import React from "react";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { validateYupSchema } from "formik";
import SignUpModal from ".";

describe("SignUpModal", () => {
  const mockState = {
    popup: {
      isSignUpOpen: false,
      /* When we pass "isEditProfileOpen, the test runs correctly, but it doesnt work with "isSignUpOpen" */

      //   But when we pass false as the value, it works for some reason
    },
  };

  const mockStore = configureStore();
  let store;
  let tree;

  beforeEach(() => {
    store = mockStore(mockState);
    tree = renderer
      .create(
        <Provider store={store}>
          <SignUpModal />
        </Provider>
      )
      .toJSON();
  });

  it("renders SignUpModal correctly", () => {
    expect(tree).toMatchSnapshot();
  });
});
