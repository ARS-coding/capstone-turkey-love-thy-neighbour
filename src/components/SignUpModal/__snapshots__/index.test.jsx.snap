// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpModal renders SignUpModal correctly 1`] = `
Array [
  <div
    className="fade modal-backdrop show"
    onClick={[Function]}
  />,
  <div
    aria-modal={true}
    className="fade modal show"
    onClick={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={Object {}}
    tabIndex={-1}
  >
    <div
      className="modal-dialog"
      id="sign-up-modal"
      onMouseDown={[Function]}
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <svg>
            logoGrayBg.svg
          </svg>
          <h2
            className="mx-auto"
          >
            Sign Up
          </h2>
          <button
            aria-label="Close"
            className="btn-close btn btn-primary"
            data-toggle="modal"
            disabled={false}
            onClick={[Function]}
            type="button"
          />
        </div>
        <div
          className="modal-body"
        >
          <div
            className="card"
          >
            <div
              className="card-body"
            >
              <form
                id="sign-up-form"
                onSubmit={[Function]}
              >
                <input
                  className="p-2"
                  id="firstName"
                  name="firstName"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="First Name"
                  value=""
                />
                <input
                  className="p-2"
                  id="lastName"
                  name="lastName"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="Last Name"
                  value=""
                />
                <select
                  className="p-2 "
                  id="gender"
                  name="gender"
                  onBlur={[Function]}
                  onChange={[Function]}
                  required={true}
                  value=""
                >
                  <option
                    defaultValue={true}
                    disabled={true}
                    value=""
                  >
                    Select a gender...
                  </option>
                  <option
                    value="Male"
                  >
                    Male
                  </option>
                  <option
                    value="Female"
                  >
                    Female
                  </option>
                  <option
                    value="Prefer not to say"
                  >
                    Prefer not to say
                  </option>
                </select>
                <select
                  className="p-2 flex-fill"
                  id="district"
                  name="district"
                  onBlur={[Function]}
                  onChange={[Function]}
                  required={true}
                  value=""
                >
                  <option
                    defaultValue={true}
                    disabled={true}
                    value=""
                  >
                    Select a district...
                  </option>
                  <option
                    value="ADALAR"
                  >
                    ADALAR
                  </option>
                  <option
                    value="ARNAVUTKÖY"
                  >
                    ARNAVUTKÖY
                  </option>
                  <option
                    value="ATAŞEHİR"
                  >
                    ATAŞEHİR
                  </option>
                  <option
                    value="AVCILAR"
                  >
                    AVCILAR
                  </option>
                  <option
                    value="BAĞCILAR"
                  >
                    BAĞCILAR
                  </option>
                  <option
                    value="BAHÇELİEVLER"
                  >
                    BAHÇELİEVLER
                  </option>
                  <option
                    value="BAKIRKÖY"
                  >
                    BAKIRKÖY
                  </option>
                  <option
                    value="BAŞAKŞEHİR"
                  >
                    BAŞAKŞEHİR
                  </option>
                  <option
                    value="BAYRAMPAŞA"
                  >
                    BAYRAMPAŞA
                  </option>
                  <option
                    value="BEŞİKTAŞ"
                  >
                    BEŞİKTAŞ
                  </option>
                  <option
                    value="BEYKOZ"
                  >
                    BEYKOZ
                  </option>
                  <option
                    value="BEYLİKDÜZÜ"
                  >
                    BEYLİKDÜZÜ
                  </option>
                  <option
                    value="BEYOĞLU"
                  >
                    BEYOĞLU
                  </option>
                  <option
                    value="BÜYÜKÇEKMECE"
                  >
                    BÜYÜKÇEKMECE
                  </option>
                  <option
                    value="ÇATALCA"
                  >
                    ÇATALCA
                  </option>
                  <option
                    value="ÇEKMEKÖY"
                  >
                    ÇEKMEKÖY
                  </option>
                  <option
                    value="ESENLER"
                  >
                    ESENLER
                  </option>
                  <option
                    value="ESENYURT"
                  >
                    ESENYURT
                  </option>
                  <option
                    value="EYÜPSULTAN"
                  >
                    EYÜPSULTAN
                  </option>
                  <option
                    value="FATİH"
                  >
                    FATİH
                  </option>
                  <option
                    value="GAZİOSMANPAŞA"
                  >
                    GAZİOSMANPAŞA
                  </option>
                  <option
                    value="GÜNGÖREN"
                  >
                    GÜNGÖREN
                  </option>
                  <option
                    value="KADIKÖY"
                  >
                    KADIKÖY
                  </option>
                  <option
                    value="KAĞITHANE"
                  >
                    KAĞITHANE
                  </option>
                  <option
                    value="KARTAL"
                  >
                    KARTAL
                  </option>
                  <option
                    value="KÜÇÜKÇEKMECE"
                  >
                    KÜÇÜKÇEKMECE
                  </option>
                  <option
                    value="MALTEPE"
                  >
                    MALTEPE
                  </option>
                  <option
                    value="PENDİK"
                  >
                    PENDİK
                  </option>
                  <option
                    value="SANCAKTEPE"
                  >
                    SANCAKTEPE
                  </option>
                  <option
                    value="SARIYER"
                  >
                    SARIYER
                  </option>
                  <option
                    value="SİLİVRİ"
                  >
                    SİLİVRİ
                  </option>
                  <option
                    value="SULTANBEYLİ"
                  >
                    SULTANBEYLİ
                  </option>
                  <option
                    value="SULTANGAZİ"
                  >
                    SULTANGAZİ
                  </option>
                  <option
                    value="ŞİLE"
                  >
                    ŞİLE
                  </option>
                  <option
                    value="ŞİŞLİ"
                  >
                    ŞİŞLİ
                  </option>
                  <option
                    value="TUZLA"
                  >
                    TUZLA
                  </option>
                  <option
                    value="ÜMRANİYE"
                  >
                    ÜMRANİYE
                  </option>
                  <option
                    value="ÜSKÜDAR"
                  >
                    ÜSKÜDAR
                  </option>
                  <option
                    value="ZEYTİNBURNU"
                  >
                    ZEYTİNBURNU
                  </option>
                </select>
                <input
                  className="p-2"
                  id="email"
                  name="email"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="Email"
                  type="email"
                  value=""
                />
                <input
                  className="p-2"
                  id="password"
                  name="password"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="Password"
                  type="password"
                  value=""
                />
                <input
                  className="p-2"
                  id="repeatedPassword"
                  name="repeatedPassword"
                  onBlur={[Function]}
                  onChange={[Function]}
                  placeholder="Repeat Password"
                  type="password"
                  value=""
                />
              </form>
            </div>
          </div>
        </div>
        <div
          className="first-sign-up-modal-footer d-flex flex-column align-items-stretch modal-footer"
        >
          <button
            className="sc-bdnxRM sc-jSFjdj hPJaof kySEpq"
            disabled={false}
            form="sign-up-form"
            type="submit"
          >
            Sign Up
          </button>
          <button
            className="sc-bdnxRM sc-gKAaRy hPJaof gsgZIm"
            onClick={[Function]}
            type="submit"
          >
            Sign Up With Google
          </button>
          <button
            className="sc-bdnxRM sc-iCoGMd hPJaof jgJmMT"
            onClick={[Function]}
            type="submit"
          >
            Sign Up With Facebook
          </button>
        </div>
        <div
          className="second-sign-up-modal-footer d-flex flex-column align-items-center modal-footer"
        >
          <span>
            Already got an
             
            <a
              href="/"
              onClick={[Function]}
            >
              Account
            </a>
            ?
          </span>
        </div>
      </div>
    </div>
  </div>,
]
`;
